pilot/pkg/proxy/envoy/v1/watcher.go

```diff
@@ -211,22 +224,27 @@ func watchCerts(ctx context.Context, certsDirs []string, watchFileEventsFn watch
 	watchFileEventsFn(ctx, fw.Event, minDelay, updateFunc)
 }
 
-func generateCertHash(h hash.Hash, certsDir string, files []string) {
+// generateCertHash generates the hash value based on the contents of cert files in the certs directory.
+// It returns false if requireFiles is true and any of the files does not exist.
+func generateCertHash(h hash.Hash, certsDir string, files []string, requireFiles bool) bool {
```

Comment:
Lets make this return an `error`, `bool` doesn't really match the semantics of what we're doing. I think it'll also simplify the logic here a bit:
```go
if _, err := os.Stat(certsDir); os.IsNotExist(err) && requiredFiles {
    return err
}
for _, file := range files {
    if bs, err := iouril.ReadFile(filename); err != nil && requiredFiles {
        return err
    } else {
        _ = h.Write(bs)
    }
}
```
Also, you can ignore the error from hash.Write: the SHA256 and FNV impls will never return an error.

Reply:
Actually, the code snippet I posted isn't _quite_ correct, actually, since we'll still try to write the hash when `err != nil` and `!requiredFiles`. However, in that case `bs` will be `nil` and not affect the running SHA calculation. Might be worth testing; otherwise just use the condition body you already have.

Reply:
Tested your approach works :)